Support migration windows for BIND9 (#1428)
openstack-k8s-operators/data-plane-adoption: docs_user/modules/proc_adopting-the-dns-service.adoc

[role="_abstract"]
To adopt the {dns_first_ref}, you patch an existing `OpenStackControlPlane` custom resource (CR) where the {dns_service} is disabled. The patch starts the service with the configuration parameters that are provided by the {rhos_prev_long} ({OpenStackShort}) environment.

[NOTE]
====
If your {rhos_prev_long} deployment uses BIND9 servers that you want to keep serving DNS during and after adoption, you can include them as external backends alongside the new operator-managed BIND9 instances. In this staged migration model, both the legacy BIND9 servers and the new cluster-hosted BIND9 StatefulSet serve the same pool concurrently. This avoids a disruptive DNS switchover and allows external DNS registries and delegations to be updated to reference the new nameservers at a later date.

After completing the standard adoption procedure, see xref:maintaining-a-migration-window-with-legacy-bind9-servers_{context}[Maintaining a migration window with legacy BIND9 servers] for the additional steps required to enable this transition window.

> ====

Reviewer: ```suggestion
If your {rhos_prev_long} deployment uses BIND9 servers that you want to serve DNS during and after adoption, you can include them as external back ends with the new operator-managed BIND9 instances. In this staged migration model, both the legacy BIND9 servers and the new cluster-hosted BIND9 StatefulSet serve the same pool concurrently. This avoids a disruptive DNS switchover and allows external DNS registries and delegations to be updated to reference the new nameservers later.

After completing the standard adoption procedure, see xref:maintaining-a-migration-window-with-legacy-bind9-servers_{context}[Maintaining a migration window with legacy BIND9 servers] for the additional steps required to enable this transition window.
====
```

----
$ oc wait --for condition=Ready --timeout=600s designate.designate.openstack.org/designate
----

[id="maintaining-a-migration-window-with-legacy-bind9-servers_{context}"]

> == Maintaining a migration window with legacy BIND9 servers

Reviewer: This procedure should be placed in its own module, not in the procedure as a sub-heading. It should follow this procedure template: https://github.com/redhat-documentation/modular-docs/blob/main/modular-docs-manual/files/TEMPLATE_PROCEDURE_doing-one-procedure.adoc?plain=1

If you want, you could merge this PR after it is approved, and I can restructure it in a separate PR. However I think this is the reason the `rdoproject.org/github-check` is failing.

Author: I had a go at restructuring.

> 

Reviewer: ```suggestion
[role="_abstract"]
```

> If your {rhos_prev_long} deployment uses BIND9 servers that should continue serving DNS during adoption, follow this procedure after completing the standard {dns_service} adoption. The designate worker sends RNDC commands and DNS NOTIFY messages to the legacy servers, which pull zone data from the mDNS service through AXFR on port 5354.

Reviewer: Can we explain a little more about why the customer would do this procedure? For example, does maintaining a migration window reduce downtime on the servers or allow for troubleshooting without disrupting workloads?

> When external DNS registries and delegations have been updated to reference the new nameservers, follow the steps in xref:completing-the-bind9-migration_{context}[Completing the BIND9 migration] to remove the legacy servers from the pool.

Reviewer: ```suggestion
```
I moved this paragraph to a new "Next steps" section at the end of the procedure.

.Procedure

> . Extract RNDC credentials from each legacy {rhos_prev_long} controller that runs BIND9 and create the `designate-external-binds` secret. Run this script from the adoption host:

Reviewer: ```suggestion
. Extract RNDC credentials from each legacy {rhos_prev_long} Controller that runs BIND9 and create the `designate-external-binds` secret. Run the following script from the adoption host:
```

      RNDC_KEYNAME=$(${!SSH_CMD} sudo grep '^key' /var/lib/config-data/ansible-generated/designate/etc/named/rndc.conf | awk '{gsub(/"/, "", $2); print $2}' | cut -f 1)
      RNDC_ALGORITHM=$(${!SSH_CMD} sudo grep algorithm /var/lib/config-data/ansible-generated/designate/etc/named/rndc.conf | awk '{gsub(/;/, "", $2); print $2}')
      CONTROLLER_IP=$(${!SSH_CMD} sudo grep 'internal_api\"' /etc/puppet/hieradata/net_ip_map.json | awk '{gsub(/"/, "", $2); gsub(/,/, "", $2); print $2}')
      cat << BINDATACR >> "$binddata_file"
    - name: controller-$i
      address: $CONTROLLER_IP
      rndcsecret: $RNDC_SECRET
      rndckeyname: $RNDC_KEYNAME
      rndcalgorithm: $RNDC_ALGORITHM
BINDATACR
    fi
  done
$ if [ -s "$binddata_file" ]; then
    external_binds_secret_file=/tmp/external_binds_secret.yaml
    cat << EOSECRETC > "$external_binds_secret_file"
apiVersion: v1
kind: Secret
metadata:
  name: designate-external-binds
  namespace: openstack
type: Opaque
stringData:
  default: |
$(cat "$binddata_file")
EOSECRETC
    oc apply -f "$external_binds_secret_file"
  fi
----
+

> where `CONTROLLER1_SSH`, `CONTROLLER2_SSH`, and `CONTROLLER3_SSH` are the SSH connection variables for each legacy controller that runs BIND9. The script extracts credentials from each available controller and automatically creates and applies the `designate-external-binds` secret in the `openstack` namespace.

Reviewer: ```suggestion
+
where:
 `CONTROLLER1_SSH`, `CONTROLLER2_SSH`, and `CONTROLLER3_SSH`::
Specifies the SSH connection variables for each legacy Controller that runs BIND9. The script extracts credentials from each available Controller and automatically creates and applies the `designate-external-binds` secret in the `openstack` namespace.
```
I don't see these variables exactly as written in the code block. That might confuse customers.

> . Reconfigure the legacy BIND9 instances to accept RNDC connections and zone transfers from the new {rhos_long} deployment. On each legacy controller, modify the BIND9 configuration. Because the `options` and `controls` blocks cannot be reopened, use the following commands to add the required configuration in-place:

Reviewer: ```suggestion
. Reconfigure the legacy BIND9 instances to accept RNDC connections and zone transfers from the new {rhos_long} deployment. On each legacy Controller, modify the BIND9 configuration. Because the `options` and `controls` blocks cannot be reopened, use the following commands to add the required configuration in-place:
```

      ${!SSH_CMD} cp -n /var/lib/config-data/ansible-generated/designate/etc/named/rndc.conf /var/lib/config-data/ansible-generated/designate/etc/named/rndc.conf.pre-adoption-backup
      ${!SSH_CMD} "sed -i 's:inet [0-9.]* allow { [0-9./]*; }:inet ${CONTROLLER_IP} allow { ${INTERNALAPI_CIDR}; }:' /var/lib/config-data/ansible-generated/designate/etc/named/rndc.conf"
      ${!SSH_CMD} cp -n /var/lib/config-data/ansible-generated/designate/etc/named/options.conf /var/lib/config-data/ansible-generated/designate/etc/named/options.conf.pre-adoption-backup
      ${!SSH_CMD} "sed -i 's:allow.notify.*};:allow-notify { $INTERNALAPI_CIDR; };:' /var/lib/config-data/ansible-generated/designate/etc/named/options.conf"
      ${!SSH_CMD} "sed -i '/allow.transfer:/d' /var/lib/config-data/ansible-generated/designate/etc/named/options.conf"
      ${!SSH_CMD} "sed -i '/^options {/,/^};/{
    /^};/ i\\\tallow-transfer { '"$INTERNALAPI_CIDR"'; };
  }' /var/lib/config-data/ansible-generated/designate/etc/named/options.conf"
  done
----
+
where:

`CONTROLLER_IP`::

> Specifies the internalapi IP address of the legacy controller. This can be retrieved from `/etc/puppet/hieradata/net_ip_map.json` on each controller.

Reviewer: ```suggestion
Specifies the internalapi IP address of the legacy Controller. You can be retrieve this from `/etc/puppet/hieradata/net_ip_map.json` on each Controller.
```

`internalapi_subnet`::

> Specifies the internalapi subnet in CIDR notation, for example `172.17.0.0/24`.

Reviewer: ```suggestion
Specifies the `internalapi` subnet in CIDR notation, for example `172.17.0.0/24`.
```

This is a CIDR range representing the network from which RHOSO designate worker

> pods send requests, not individual IP addresses.

Reviewer: ```suggestion
This is a CIDR range represents the network from which RHOSO designate worker
pods send requests.
```

+
[NOTE]
====

The `sed` commands in the shell script example are directed at default based deployments. If customizations have been made to the 
deployment, please either modify these commands or perform the modifications directly.

> 

Reviewer: ```suggestion
====
The `sed` commands in the shell script example are directed at default deployments. If customizations are made to the deployment, either modify these commands or perform the modifications directly.
```

====

+
Reload the BIND9 configuration:
+
----
$ systemctl restart desginate-designatebackendbind9
----
+
[NOTE]
====

Direct edits to the BIND9 configuration are appropriate because the legacy
control plane ({OpenStackPreviousInstaller}) is being decommissioned and will
no longer manage these files.

> ====

Reviewer: ```suggestion
====
You can edit the BIND9 configuration directly because the legacy {OpenStackPreviousInstaller} control plane is being decommissioned and no longer manages these files.
====
```

> . Open the firewall on each legacy controller to allow RNDC and DNS traffic from the {rhos_long} pods:

Reviewer: ```suggestion
. Open the firewall on each legacy Controller to allow RNDC and DNS traffic from the {rhos_long} pods:
```

+
[subs="+quotes"]
----
$ firewall-cmd --permanent --add-rich-rule='rule family="ipv4" source address="<internalapi_subnet>" port port="953" protocol="tcp" accept'
$ firewall-cmd --permanent --add-rich-rule='rule family="ipv4" source address="<internalapi_subnet>" port port="53" protocol="tcp" accept'
$ firewall-cmd --permanent --add-rich-rule='rule family="ipv4" source address="<internalapi_subnet>" port port="53" protocol="udp" accept'
$ firewall-cmd --reload
----
+
where:

`<internalapi_subnet>`::

> Specifies the internalapi subnet in CIDR notation, for example `172.17.0.0/24`.

Reviewer: ```suggestion
```
Since you already defined the variable in an earlier step, you don't have to define it again within the same procedure.

. Create a LoadBalancer service to expose the mDNS service on the internalapi

>   network so that the legacy BIND9 servers can perform AXFR zone transfers: +

Reviewer: ```suggestion
. Create a LoadBalancer service to expose the mDNS service on the internalapi
  network so that the legacy BIND9 servers can perform AXFR zone transfers: 
  +
```

----
$ cat << EOF_CAT > /tmp/designate-external-mdns-svc.yaml
apiVersion: v1
kind: Service
metadata:
  namespace: openstack
  name: designate-external-master
  labels:
    service: designate-mdns
    designate.openstack.org/external_master: "true"
  annotations:
    core.openstack.org/ingress_create: "false"
    metallb.io/ip-allocated-from-pool: internalapi
    metallb.universe.tf/address-pool: internalapi
    metallb.universe.tf/allow-shared-ip: internalapi
    metallb.universe.tf/loadBalancerIPs: 172.17.0.80
    designate.openstack.org/external_pool: default
spec:
  selector:
    service: designate-mdns
  ports:
    - name: dns-tcp
      port: 5354
      protocol: TCP
      targetPort: 5354
    - name: dns
      port: 5354
      protocol: UDP
      targetPort: 5354
  ipFamilies:
    - IPv4
  ipFamilyPolicy: SingleStack
  externalTrafficPolicy: Cluster
  internalTrafficPolicy: Cluster
  type: LoadBalancer
EOF_CAT
$ oc apply -f /tmp/designate-external-mdns-svc.yaml
----

> . Patch the {dns_service} CR to reference the external binds secret:

Reviewer: ```suggestion
. Patch the {dns_service} custom resource to reference the external binds secret:
```

+
----
$ oc patch openstackcontrolplane openstack --type=merge --patch '
spec:
  designate:
    template:
      externalBindsSecret: designate-external-binds
'
----

. Wait for the {dns_service} to reconcile:
+
----
$ oc wait --for condition=Ready --timeout=300s designate.designate.openstack.org/designate
----

. Verify that the external BIND9 targets are included in the pool configuration:
+
----
$ oc get configmap designate-pools-yaml-config-map -o jsonpath='{.data.pools\.yaml}' | grep -A5 'external-bind9'
----
+
The output should list the external BIND9 servers you configured in the external binds secret.

. Verify RNDC connectivity from the designate worker pods to the legacy BIND9 servers:
+
[subs="+quotes"]
----
$ oc exec -it $(oc get pods -l service=designate-worker -o name | head -1) -- \
    rndc -s <controller_internalapi_ip> -k /etc/designate/rndc-keys/default-rndc-0 status
----
+
where:

`<controller_internalapi_ip>`::
Specifies the internalapi IP address of a legacy BIND9 server.
+
A successful response indicates that the designate worker can manage the external BIND9 server.

> 

Reviewer: ```suggestion

.Next steps
When external DNS registries and delegations are updated to reference the new nameservers, remove the legacy servers from the pool. For more information, see xref:completing-the-bind9-migration_{context}[Completing the BIND9 migration].
```

[id="completing-the-bind9-migration_{context}"]

> == Completing the BIND9 migration

Reviewer: Same comment as above. This procedure should be placed in its own module, not in the procedure as a sub-heading. It should follow this procedure template: https://github.com/redhat-documentation/modular-docs/blob/main/modular-docs-manual/files/TEMPLATE_PROCEDURE_doing-one-procedure.adoc?plain=1

After external DNS registries and delegations have been updated to reference
the new {rhos_long} nameservers, remove the legacy BIND9 servers from the

> {dns_service} pool.

Reviewer: ```suggestion
After external DNS registries and delegations have been updated to reference
the new {rhos_long} nameservers, remove the legacy BIND9 servers from the
{dns_service} pool to complete the BIND9 migration.
```

.Procedure

> . Remove the external binds secret reference from the {dns_service} CR:

Reviewer: ```suggestion
. Remove the external binds secret reference from the {dns_service} custom resource:
```
